Sources/Subs-Auth.php

```diff
@@ -943,4 +950,60 @@ function hash_benchmark($hashTime = 0.2)
 	return $cost;
 }
 
+if (!function_exists('hash_pbkdf2'))
+{
+	function hash_pbkdf2($algo, $password, $salt, $iterations, $length = 0, $raw_output = false)
+	{
+		if (!in_array(strtolower($algo), hash_algos()))
+			trigger_error(__FUNCTION__ . '(): Unknown hashing algorithm: ' . $algo, E_USER_WARNING);
+		if (!is_numeric($iterations))
+			trigger_error(__FUNCTION__ . '(): expects parameter 4 to be long, ' . gettype($iterations) . ' given', E_USER_WARNING);
+		if (!is_numeric($length))
+			trigger_error(__FUNCTION__ . '(): expects parameter 5 to be long, ' . gettype($length) . ' given', E_USER_WARNING);
+		if ($iterations <= 0)
+			trigger_error(__FUNCTION__ . '(): Iterations must be a positive integer: ' . $iterations, E_USER_WARNING);
+		if ($length < 0)
+			trigger_error(__FUNCTION__ . '(): Length must be greater than or equal to 0: ' . $length, E_USER_WARNING);
+
+		$output = '';
+		$block_count = $length ? ceil($length / strlen(hash($algo, '', $raw_output))) : 1;
+		for ($i = 1; $i <= $block_count; $i++)
+		{
+			$last = $xorsum = hash_hmac($algo, $salt . pack('N', $i), $password, true);
+			for ($j = 1; $j < $iterations; $j++)
+			{
+			    $xorsum ^= ($last = hash_hmac($algo, $last, $password, true));
+			}
+			$output .= $xorsum;
+		}
+
+		if (!$raw_output)
+			$output = bin2hex($output);
+		return $length ? substr($output, 0, $length) : $output;
+	}
+}
+
+if (!function_exists('hash_equals'))
+{
+	function hash_equals($known_string, $user_string)
```

Comment:
I see this came from the php.net/hash_equals user comments.  Do we need to credit that or is that free use?

Reply:
Good question. I'll check.

Reply:
According to the comment submission page, "any note submitted here becomes the property of the PHP Documentation Group and will be available under the same license as the documentation." The PHP Documentation is licensed using the Creative Commons Attribution 3.0 License. So I guess the answer is yes, we do need to. I'll add a comment doing so.